### PR TITLE
Trace return values too

### PR DESCRIPTION
**Pull request by encounter**
Before, dllexport functions were generated like:
```rust
fn name(args) -> ret {
  trace(args);
  // ... user code
}
```
Now they look like:
```rust
fn name(args) -> ret {
  fn name_impl(args) -> ret {
    // ... user code
  }
  let ctx = trace_begin(args);
  let ret = name_impl(args);
  trace_return(ctx, ret);
  ret
}
```
The idea is to format the arguments _before_ running the impl, in case the impl modifies them. Then the log line is printed after the impl returns and includes the return value.

One interesting consequence of this approach is that dllexport impl arguments now are actually checked for unused-ness, since they're not considered used by the `trace` call. So I had to add `allow(unused_variables)` to a bunch of files. I think this is a good thing, though, and could later remove that `allow` when cleaning up files and renaming unused parameters to have a `_` prefix. I didn't do that in this PR, since that would be a massive changeset.

-----

One unsolved problem is with implementations calling other dllexport functions (for example, `ReadFile` 
calling `SetLastError`), `SetLastError`'s trace line will be printed before `ReadFile`'s trace line. One idea is to swap it around so it looks like:
```rust
fn name(args) -> ret {
  // ... user code
}
fn name_trace(args) -> ret {
  let ctx = trace_begin(args);
  let ret = name(args);
  trace_return(ctx, ret);
  ret
}
```
Then `builtin` would call the `name_trace` function instead. This way we can internally call other winapi functions without creating more trace logs. I tried this originally, but it required odd changes where modules that did `pub use module::InnerDllExport` to do `pub use module::InnerDllExport_trace`, which felt wrong. So I thought I'd submit it in its current state and revisit that later if you like this idea in general.

### CODE REVIEW

File: win32/derive/src/trace.rs [hunk 1]
```diff
@@ -1,34 +1,75 @@
-use quote::quote;
+use quote::{format_ident, quote};
 
 /// Insert a call to trace::trace() at the front of a function that logs its name and arguments.
```
Comment (evmar) [inline win32/derive/src/trace.rs:3]: This comment now out of date

File: win32/src/winapi/advapi32.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#![allow(non_snake_case)]
+#![allow(non_snake_case, unused_variables)]
```
Comment (evmar) [inline win32/src/winapi/advapi32.rs:1]: Why are these needed?

I think the macro can add the suppressions on each dllexport.  That would mean non-exported (non-traced) functions still get unused variable warnings.

File: win32/derive/src/trace.rs [hunk 3]
```diff
@@ -1,34 +1,75 @@
...
+                    pat.by_ref = None;
+                    pat.mutability = None;
+                }
+                _ => {}
+            },
+            _ => {}
+        };
+    }
+
+    let impl_name = format_ident!("{}_impl", name);
```
Comment (evmar) [inline win32/derive/src/trace.rs:59]: I think you can avoid needing to generate a name here by generating code like

```
#prolog
let ret = (|arg1, arg2| { #func })(arg1, arg);
#epilog
```

I think what you have is fine too, just suggesting it in case you like it more

**Review by evmar — COMMENTED** [review]
I love this!

One thing I wonder about is that this means we no longer trace upon entering functions, only when completing them, which means that if we e.g. crash within a function the log won't show it.  In particular for async functions it can be a long time before they complete.  I have no useful suggestion here, maybe it's fine.

Regarding dllexport functions calling each other, the pattern I've been doing elsewhere is that if there's functionality that is needed internally we expose it with a `rust_like` name and use that from the dllexported `WindowsLike` name. 

Example https://github.com/evmar/retrowin32/blob/0faaa1c45bb8413797f349ebd0ca13af89cd3d01/win32/src/winapi/user32/window.rs#L134C4-L134C18
for RegisterClass

File: win32/src/winapi/advapi32.rs (hunk 2 above)
Comment (encounter) [reply win32/src/winapi/advapi32.rs:1]: I mention this in the description, but it’s because the tracing made the arguments always-used before. So now hundreds of warnings get thrown. I think we should fix them instead of ignoring them, but for the sake of the PR size, opting to ignore them at a file level.

File: win32/derive/src/trace.rs (hunk 3 above)
Comment (encounter) [reply win32/derive/src/trace.rs:59]: I did it this way so that inner `return`s don’t return the outer fn. I don’t remember if closures would work the same, tbh

**Comment by evmar** [comment]
I merged this with two small changes:
1) moved the allow-unused to the function scope
2) made the return values trace as hex, for consistency with how we print all numbers in traces
